### ISSUE TITLE
Segfault on server finalization when num_emerge_threads > 1

### ISSUE THREAD

**Issue by ghost** [minetest/minetest#8346]
##### Issue type

 - Bug report
 
##### Minetest version

stable-0.4, stable-5, master

##### Summary

The server randomly crashes on shutdown when using a mod which calls `get_mapgen_object` within `register_on_generated`. The backtrace has always this form:

```
#0  0x0000555555f1bd08 in std::thread::get_id (this=0x0)
    at /usr/include/c++/6/thread:175
#1  0x0000555555f1bd61 in Thread::getThreadId (this=0x555556b0a580)
    at $minetest/src/threading/thread.h:97
#2  0x0000555555f1bd27 in Thread::isCurrentThread (this=0x555556b0a580)
    at $minetest/src/threading/thread.h:92
#3  0x0000555555f9f89e in EmergeManager::getCurrentMapgen (this=0x555556b0a0b0)
    at $minetest/src/emerge.cpp:209
#4  0x0000555555e897b7 in ModApiMapgen::l_get_mapgen_object (L=0x4009e378)
    at $minetest/src/script/lua_api/l_mapgen.cpp:676
#5  0x0000555555e40005 in script_exception_wrapper (L=0x4009e378, 
    f=0x555555e896c4 <ModApiMapgen::l_get_mapgen_object(lua_State*)>)
    at $minetest/src/script/common/c_internal.cpp:38
#6  0x00007ffff4f9015c in ?? ()
   from /usr/lib/x86_64-linux-gnu/libluajit-5.1.so.2
#7  0x00007ffff4fd59e0 in lua_pcall ()
   from /usr/lib/x86_64-linux-gnu/libluajit-5.1.so.2
#8  0x0000555555e464db in ScriptApiBase::runCallbacksRaw (this=0x555556b1c0b0, 
    nargs=3, mode=RUN_CALLBACKS_MODE_FIRST, 
    fxn=0x5555561457f0 <ScriptApiEnv::environment_OnGenerated(irr::core::vector3d<short>, irr::core::vector3d<short>, unsigned int)::__FUNCTION__> "environment_OnGenerated")
    at $minetest/src/script/cpp_api/s_base.cpp:259
#9  0x0000555555e4d271 in ScriptApiEnv::environment_OnGenerated (this=
    0x555556b1c080, minp=..., maxp=..., blockseed=3615170633)
    at $minetest/src/script/cpp_api/s_env.cpp:41
#10 0x0000555555fa0e88 in EmergeThread::finishGen (this=0x555556b0aa60, 
    pos=..., bmdata=0x7fffb23fb9d0, modified_blocks=0x7fffb23fb920)
    at $minetest/src/emerge.cpp:588
#11 0x0000555555fa1559 in EmergeThread::run (this=0x555556b0aa60)
    at $minetest/src/emerge.cpp:653
#12 0x0000555555eb9d33 in Thread::threadProc (thr=0x555556b0aa60)
    at $minetest/src/threading/thread.cpp:196
#13 0x0000555555eba95e in std::_Bind_simple<void (*(Thread*))(Thread*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x7fffb8080468)
    at /usr/include/c++/6/functional:1391
#14 0x0000555555eba8af in std::_Bind_simple<void (*(Thread*))(Thread*)>::operator()() (this=0x7fffb8080468) at /usr/include/c++/6/functional:1380
#15 0x0000555555eba88e in std::thread::_State_impl<std::_Bind_simple<void (*(Thread*))(Thread*)> >::_M_run() (this=0x7fffb8080460)
    at /usr/include/c++/6/thread:197
#16 0x00007ffff3450e6f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#17 0x00007ffff48e6494 in start_thread (arg=0x7fffb23fc700)
    at pthread_create.c:333
#18 0x00007ffff2bc5acf in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

See https://github.com/minetest/minetest/pull/8328#issuecomment-470345036 for a `bt full` taken by argyle77. Most importantly, `#4` reveals the Lua function called, and `#3` is the main culprit.

I think what is happening here is a simple race. Consider this scenario:

- The mod uses a Lua mapgen which first uses `get_mapgen_object`, then generates stuff, then calls `get_mapgen_object` again, then generates more stuff. That's exactly the case with `magma_conduits`, see https://github.com/minetest-mods/magma_conduits/blob/9a21f59d296eab08daf51e29941e38d6292e8ac1/volcanoes.lua#L137.
- There are two emerge threads, 0 and 1. Thread 0 is idle; thread 1 is busy generating the first stuff.
- Server shutdown is requested. Server calls `EmergeManager::stopThreads()` which first asks all threads to terminate, then waits for each to finish in order. So, both threads are asked to terminate.
- Server waits for thread 0 to finish, which happens immediately because it was idle. When it finishes, its `m_thread_obj` field (i.e. the `std::thread` object within the MT `Thread` object) is set to `nullptr` (in `Thread::wait`).
- Meanwhile, thread 1 is still busy in the first part of the generation. The termination request isn't checked until after Lua finishes.
- When the first part of the generation finishes, `get_mapgen_object` is called. This in turn calls `EmergeManager::getCurrentMapgen()` which loops through all threads, beginning with the already stopped thread 0, asking each if `isCurrentThread()`.

The last step is fatal. `isCurrentThread()` needs to query the id of the `m_thread_obj` object, which (as shown in the traceback) is null for thread 0.

It's not difficult to imagine a scenario where bad luck makes thread 1 reach the first `get_mapgen_object` call, which would make the crash possible, albeit unlikely, in other cases.

The fix is to check if the thread `isRunning()` before checking if `isCurrentThread()` within `EmergeManager::getCurrentMapgen()`. But there's a catch. There's a possible TOCTTOU race between `isRunning` and `isCurrentThread` where the thread is deleted right after the first check. Which leads to this patch:

https://gitlab.com/pgimeno/minetest/commits/pg-fix-crash-on-exit

A simpler but hackier solution is to reverse the direction of the wait loop in the `EmergeManager::stopThreads` method (or in the `getCurrentMapgen` method, but not in both). With that change, `isCurrentThread` won't check a destroyed thread unless it's called from a thread that is not an emerge thread, which I don't think can happen during shutdown.

**Comment by paramat** [comment]
When you make a patch on Gitlab, how do i 'get' the patch in '.patch' form to create a Github PR from it.

**Comment by paramat** [comment]
How simple and safe is your patch? I'm not sure about merging it for 5.0.1, especially since it's only a problem if emerge threads > 1 and that already causes other bugs.

**Comment by pgimeno** [comment]
> When you make a patch on Gitlab, how do i 'get' the patch in '.patch' form to create a Github PR from it?

In this case, since it's a single commit, you can click on the commit, then edit the URL to add .patch or .diff at the end.

But really, git excels in being able of working with a network of repositories, not just with one. It's more straightforward than it may seem to add someone's repository as a remote and fetch a particular branch from it which you can check out like any other branch; I've posted the commands several times. Getting used to do fetch + merge instead of pull also helps, because it's the same for origin or upstream as for any other remote or branch.

```
# Add a remote repository with the name pgimeno (this needs to be done only once)
git remote add pgimeno https://gitlab.com/pgimeno/minetest.git
# Fetch the pg-fix-crash-on-exit branch from that remote
git fetch pgimeno pg-fix-crash-on-exit
# Create and check out a local branch for it
git checkout pg-fix-crash-on-exit
```

Then if you want to remove everything again:
```
# Switch to a different branch
git checkout master
# Remove the local branch
git branch -D pg-fix-crash-on-exit
# Remove the entire remote with all its branches (not really necessary if you're going to work with it)
git remote remove pgimeno
```

> How simple and safe is your patch? I'm not sure about merging it for 5.0.1, especially since it's only a problem if emerge threads > 1 and that already causes other bugs.

Agreed, it's probably not worth adding it for 5.0.1 given the other problem. However it's quite straightforward; it's just an additional check with some simple locking added, which affects very small sections that don't wait for anything else, therefore I think that there's no possible deadlock.

**Comment by paramat** [comment]
This might be fixed now due to a recent commit.

**Comment by sfan5** [comment]
This is the exact bug fixed by cb159f8d8af4556391db8a6875657625733eeb11.
